### PR TITLE
Security Hardening

### PR DESCRIPTION
**Pull request by matusik-ops**

### CODE REVIEW

File: Terraform/emty_folde_for_terraform.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+this is empty folde for terraform
```
Comment (matofeder) [inline Terraform/emty_folde_for_terraform.txt]: an alternative solution could be: 
Create an empty file called .gitkeep in the directory, and git add it.

File: Vagrant/.vagrant/machines/master/virtualbox/action_provision [hunk 2]
```diff
@@ -0,0 +1 @@
+1.5:4d28ffe5-0e15-49f7-9577-e3516e389471
```
Comment (matofeder) [inline Vagrant/.vagrant/machines/master/virtualbox/action_provision]: The whole .vagrant directory contains your local vagrant state and it is not intended to be here in github. 
Remove .vagrant dir and just keep `Vagrantfile ` here

File: Ansible/playbook.yml [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  tasks:
+
+    #- name: Test connection
+    #  local_action: command ssh -q -o BatchMode=yes -o ConnectTimeout=3 {{ inventory_hostname }}
+    #  register: test_user
+    #  ignore_errors: true
+    #  changed_when: false
+
+    - name: Update and Upgrade apt packages
+      apt:
```
Comment (matofeder) [inline Ansible/playbook.yml]: A good practice is to utilize built-in modules as follows `ansible.builtin.apt`.

Also, ansible-lint will be happy then, see e.g. https://www.redhat.com/sysadmin/ansible-lint

File: Vagrant/Vagrantfile
Comment (chess-knight) [inline Vagrant/Vagrantfile:1]: Please, add the `.vagrant` folder to the `.gitignore`. These files should not be stored in Git!

File: Ansible/playbook.yml [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        update_cache: yes
+        upgrade: dist
+
+    - name: Create a non-root user
```
Comment (matofeder) [inline Ansible/playbook.yml]: IMO name is confusing, consider name like `Create a new admin user`

File: Ansible/playbook.yml [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      user:
+        name: "{{ new_user }}"
+        password: "{{ new_user_password | password_hash('sha512') }}"
```
Comment (matofeder) [inline Ansible/playbook.yml]: A good idea is to use an SSH key instead of a password.

File: Ansible/playbook.yml [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        line: 'PasswordAuthentication no'
+          #notify: restart ssh
+  
+    #- name: Allow specific users for SSH
+    #  lineinfile:
+    #    path: /etc/ssh/sshd_config
+    #    line: 'AllowUsers {{ new_user }}'
+    #  notify: restart ssh
+        
+    - name: Enable UFW and allow SSH
```
Comment (chess-knight) [inline Ansible/playbook.yml]: I can see this is not only `SSH`, but also HTTP(S)!

File: README.md
Comment (chess-knight) [inline README.md:1]: IMO, README is best practice, at least write some intro and commands on how to run it.

File: Ansible/playbook.yml [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        update_cache: yes
+        upgrade: dist
+
+    - name: Create a non-root user
+      user:
+        name: "{{ new_user }}"
+        password: "{{ new_user_password | password_hash('sha512') }}"
+        groups: sudo
+        shell: /bin/bash
+        state: present
```
Comment (matofeder) [inline Ansible/playbook.yml:23]: Also a good idea could be the following task 
```
- name: Allow user {{ new_user }} to run all commands without password
  ansible.builtin.copy:
    dest: "/etc/sudoers.d/{{ new_user }}"
    content: "{{ new_user }} ALL=(ALL) NOPASSWD: ALL\n"
    owner: root
    group: root
    mode: '0440'
```

File: Ansible/inventory [hunk 8]
```diff
@@ -0,0 +1,10 @@
+all:
+  hosts:
+    vm2:
+      ansible_host: 192.168.56.12
+      ansible_user: vagrant
+      ansible_ssh_pass: vagrant
+      ansible_ssh_private_key_file: ~/.ssh/id_rsa
+     # admin_user: matusik
+     # admin_ssh_pass: matusik
+     # admin_ssh_private_key_file: ~/.ssh/id_rsa
```
Comment (matofeder) [inline Ansible/inventory]: Remove all lines that are commented and unnecessary, and do not add any value to the code.
Check the whole PR

File: Ansible/playbook.yml [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        #regexp: '^#?Port '
+        #line: 'Port 2222'
+        #notify: restart ssh
+          #
+    - name: Disable password authentication for SSH
+      lineinfile:
+        path: /etc/ssh/sshd_config
+        regexp: '^#?PasswordAuthentication'
+        line: 'PasswordAuthentication no'
+          #notify: restart ssh
```
Comment (matofeder) [inline Ansible/playbook.yml]: to apply sshd changes you have to trigger the ssh handler somewhere

File: Ansible/playbook.yml [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  
+    #- name: Allow specific users for SSH
+    #  lineinfile:
+    #    path: /etc/ssh/sshd_config
+    #    line: 'AllowUsers {{ new_user }}'
+    #  notify: restart ssh
```
Comment (matofeder) [inline Ansible/playbook.yml]: also a good practice, why not apply it?

**Comment by matofeder** [comment]
The next exercise could be utilizing Github actions to execute pipelines with ansible linter and YAML linter

File: README.md [hunk 11]
```diff
@@ -1,2 +0,0 @@
-# automation
-for project: Project_myweb
```
Comment (matofeder) [inline README.md:2]: Could you add some README content?
README should contain  at least info about what is the content/purpose of this repository

File: Ansible/playbook.yml [hunk 12]
```diff
@@ -0,0 +1,136 @@
+---
+- name: Security Hardening on Ubuntu
+  hosts: all
+  gather_facts: no
+  become: yes
+  vars:
+    ssh_public_keys:
```
Comment (chess-knight) [inline Ansible/playbook.yml]: In my opinion, these variables can be better structured. I can see that `ssh_public_keys` are keys for the new_user.
Something like:
```yaml
new_user:
  name: ...
  password: ...
  ssh_public_keys: ...
```
And also ... Is the password really needed here?

File: Terraform/emty_folde_for_terraform.txt (hunk 1 above)
Comment (matusik-ops) [reply Terraform/emty_folde_for_terraform.txt]: folder or file?

File: Ansible/playbook.yml (hunk 9 above)
Comment (matusik-ops) [reply Ansible/playbook.yml]: hendlers are at the end of the code

File: Ansible/playbook.yml (hunk 10 above)
Comment (matusik-ops) [reply Ansible/playbook.yml]: yes, but repeated launch of the playbook with current user (Vagrant) wont work

File: Terraform/emty_folde_for_terraform.txt (hunk 1 above)
Comment (matofeder) [reply Terraform/emty_folde_for_terraform.txt]: file

.gitkeep is a common thing you will see in codebases, where an empty folder needs to be tracked via Git.

The name of the dummy file may not always be .gitkeep but you'll see the actual practice over and over again as a developer.

File: Ansible/playbook.yml (hunk 7 above)
Comment (matusik-ops) [reply Ansible/playbook.yml:23]: ok but file /etc/sudoers.d/{{ new_user }} does not exist during the execution of the playbook so i have to add one line force: false...that's right?

Comment (matofeder) [reply Ansible/playbook.yml:23]: Just ensure that the dir exists before the above task, e.g.

```
- name: Ensure '/etc/sudoers.d' folder exists
  file:
    dest: /etc/sudoers.d
    state: directory
    owner: root
    group: root
```

File: Vagrant/Vagrantfile
Comment (matusik-ops) [reply Vagrant/Vagrantfile:1]: i did it, but how can i remove this file .vagrant from git repository?

Comment (chess-knight) [reply Vagrant/Vagrantfile:1]: `git rm -r --cached .vagrant`

Comment (matusik-ops) [reply Vagrant/Vagrantfile:1]: Wow...I did it. It is owesome. Thank you :-)

Comment (chess-knight) [reply Vagrant/Vagrantfile:1]: You can commit that `.gitignore` file

File: README.md [hunk 13]
```diff
@@ -1,2 +1,34 @@
...
+### Ansible: core 2.16.11
+### jinja: 3.1.4
+
+# How to run it
+
+1. Install update
+2. Git clone repo "Automation" to /home directory
+```bash
+cd
+git clone ...
```
Comment (matofeder) [inline README.md]:
```suggestion
git clone https://github.com/matusik-ops/automation.git
```

File: README.md [hunk 14]
```diff
@@ -1,2 +1,34 @@
...
+### Docker-compose: 2.28.1
+### Ansible: core 2.16.11
+### jinja: 3.1.4
+
+# How to run it
+
+1. Install update
+2. Git clone repo "Automation" to /home directory
+```bash
+cd
```
Comment (matofeder) [inline README.md]:
```suggestion

```

File: README.md [hunk 15]
```diff
@@ -1,2 +1,34 @@
 # automation
-for project: Project_myweb
+for project: Project_myweb/automation
+
+### Requirements:
+
+Versions:
+### Python3: 3.10.12
```
Comment (matofeder) [inline README.md]: `###` is title in markdown, use `-` or `*` instead

File: Ansible/playbook.yml [hunk 16]
```diff
@@ -0,0 +1,108 @@
+---
+- name: Security Hardening on Ubuntu
+  hosts: all
+  gather_facts: false
```
Comment (matofeder) [inline Ansible/playbook.yml]: it is needed?

**Review by matofeder — APPROVED** [review]
LGTM, 
just minor comments

**Comment by matofeder** [comment]
@matusik-ops feel free to merge this PR

**Pull request merged**